### ISSUE TITLE
Anonym loops and multiline scripts

### ISSUE THREAD

**Issue by raingloom** [JacquesLucke/animation_nodes#598]
**TLDR: if a loop or a script is only ever used in one place, then there should be no need to separate invocation from defition and polluting a "global" namespace that can't even be searched from the UI.**

![screen](https://cloud.githubusercontent.com/assets/8085052/17840051/cbbe7db6-67fb-11e6-951f-453fd7f8ba9c.jpg)

**long:**

Right now every loop and script needs to be defined and then called by name. I don't know about everyone else, but so far I have not needed to use a single loop in more than one place, so having to name them makes moving things around a bit more difficult and the name list more cluttered.

I think it would be easier if *single-use* loops and scripts could be inserted anywhere without having to go through all of that. Since the _"defining"_ node already has input and output sockets, it would be easiest to use those. _(which is what I tried to do first, as I imagine many other beginners did)_

For scripts, it's a bit different. I'm not an expert on Blender's API but I don't think embedding a script editor within the node view would be easy, but a multiline version of the `expression` node would be an improvement already.

My biggest problem is the naming. I don't like having to look through a list for the thing I want to use, especially when I know its name and could just type it. _(yes, a node based scripting environment may not be for me)_
I think it just doesn't fit well with how other parts of the Blender UI work. Eg.: when choosing a target for a constraint the user can search the object list, not just scroll through until they find it. _(not to mention the picker that was added)_
The fact that loops can have the same name doesn't help either.

**Comment by JacquesLucke** [comment]
sorry, I don't understand what you mean. Can you maybe explain it a bit more detailed and provide an example?

**Comment by raingloom** [comment]
Yes, I'll update the original.

**Comment by JacquesLucke** [comment]
In fact I use the same loop relatively often but I'll continue to write under the assumption that a loop is only used once.

It's not easy to organize my thoughts on the topic right now but I'll try.

Mixing loops with nodes that are not in a loop is complicated, maybe even impossible. Because AN cannot know which nodes will be executed multiple times and which nodes only one time. By enforcing the separation this problem is completely gone. Also very often loops do have outputs. That doesn't really fit into the model you are proposing. (maybe you can make a little sketch how it could look like with more nodes?)

Personally I never go through the menu to insert the Invoke Subprogram node to a given loop. There is a shortcut for this. When the Loop Input (or group input or script) node is selected you can press W and move the mouse to the right. The W key also has some other interesting functionality you might not know. Just try it out yourself (eg select an Object Instancer node and press W to create a loop)

I understand that it is a bit unusual to use the search box to select a socket type. In fact, at first it was a normal drop down menu that is used in many other places in Blender. However this became unusable! There are simply to many socket types. Also the dropdown menus were a bit harder to manage because when they aren't treated very carefully by all developers they sometimes switch to another entry when elements are added/removed etc. Using the search removed this problem as well (not everywhere but in many cases)

I have to admit, I don't like the Blender search box very much. It is not very flexible when you don't type the search string exactly right. I'm thinking of ways to improve that.

You are right. Blender does not have multiline textboxes, unfortunally. That would make some things a lot easier. I had to implement multiline stuff myself in many areas (debug drawer node, problems panel, ...). I'd love to have them as well (although it would not be an expression anymore then). It is just not really possible with multiple very ugly hacks. I'm not sure if it is possible to do this with python.

Having this separation between subprograms gives me more flexibility, extensibilty, modularity, ... 
This is why I really like the subprogram "design" as it is nowadays. If there are things you do very often think about automating it.  Like I did with these operators in the W pie menu. btw E also has a pie menu. And U opens the advanced settings for the active node. 

Naming things is optional but should be done for better overview. I don't agree with your statement that all groups that are used only one time should be integrated in the main node network. Working on abstractions makes it much easier to archieve complex results.
